modules/mqnic/mqnic_netdev.c

```diff
@@ -340,15 +340,25 @@ static int mqnic_ioctl(struct net_device *ndev, struct ifreq *ifr, int cmd)
 	}
 }
 
+#define MQNIC_BASE_NETDEV_OPS \
+	.ndo_open = mqnic_open, \
+	.ndo_stop = mqnic_close, \
+	.ndo_start_xmit = mqnic_start_xmit, \
+	.ndo_get_stats64 = mqnic_get_stats64, \
+	.ndo_validate_addr = eth_validate_addr, \
+	.ndo_change_mtu = mqnic_change_mtu, \
+	.ndo_do_ioctl = mqnic_ioctl
+
 static const struct net_device_ops mqnic_netdev_ops = {
-	.ndo_open = mqnic_open,
-	.ndo_stop = mqnic_close,
-	.ndo_start_xmit = mqnic_start_xmit,
-	.ndo_get_stats64 = mqnic_get_stats64,
-	.ndo_validate_addr = eth_validate_addr,
-	.ndo_change_mtu = mqnic_change_mtu,
-	.ndo_do_ioctl = mqnic_ioctl,
+	MQNIC_BASE_NETDEV_OPS,
+};
+
+#ifdef CONFIG_MQNIC_SELQ_HANDLER
+static const struct net_device_ops mqnic_selq_netdev_ops = {
```

Comment:
My only concern about structure swapping like this for a single feature is what happens if we add a few more features down the road that need this, and then you end up with 2^n structures.  Maybe we should just make a per-instance copy of `net_device_ops` somewhere and then replace just the pointers that need to be changed.  Perhaps just add a `struct net_device_ops` entry to `struct mqnic_dev`, init that from `mqnic_netdev_ops` and then just directly overwrite `ndo_select_queue` as necessary?

Reply:
Sounds good, understood:

So (1) I'll get rid of the compile-time option and (2) we'll just have the module parameter, which decides about enabling the selq API/feature. However (3) the parameter would be "disabled" then by default. Enabling it on module load time would then, as you said, just put the pointer to the select_queue function into the copied struct.


Thinking ahead + loud about our ZynqMP port, where we will have 2 classes of interfaces, where just one of them needs selq support and since performance is a different/more relevant story there of course (CPU-wise), I also came to think if the module parameter actually should be an array of enable/disable switches to really control that feature per interface. OTOH module parameter arrays might be a little bit odd to some ...
As soon as the mqnic driver is dynamically registering (Linux) network interfaces (via `rtnetlink`) based on custom (mqnic) interface/port(s) "combinations", enabling selq could be made an argument for a `newlink()` call and thus per netdev. But that's in the more distant future, I guess.